### DISCUSSION TITLE
WinUI SwapChain

### DISCUSSION THREAD

**Discussion by TomAtWork** [amerkoleci/Vortice.Windows#164] [Q&A] [answered]
Hello,
i am trying to test Vortice.Windows together with WinUI (compare https://github.com/amerkoleci/Vortice.Windows/issues/91). It seems the WinUI functionality was removed or how can i cast to a ISwapChainPanelNative that works with WinUI?

Thanks in advance,
Tom

**Comment by amerkoleci** [comment, marked as answer]
There were some issues with SharpGen and WinUI, probably will bring it back sometime but not soon probably.

**Reply by TomAtWork** [reply]
Thanks for the info,
you are referring to this [issue](https://github.com/SharpGenTools/SharpGenTools/issues/181), right ?
It is my understanding that this just because of the namespaces, and currently Vortice.Windows supports UWP SwapChainPanels and not WinUI SwapChainPanels and we cannot have both ?
I'm currently investigating to migrate a library using

DirectX11 / .Net 4.8 / WinForms / WPF / SharpDX 
to 
DirectX11 /.NET 6 WinForms / WPF / WinUI and hopefully Vortice.Windows.

If i replace the reference *windows.ui.xaml.media.dxinterop.h* to *microsoft.ui.xaml.media.dxinterop.h* in *Vortice.DXGI/Mappings.xml* will this work to have a WinUI capable version of Vortice.Windows, or am i just being naive ?

Anyway, i have a preliminary solution that is working for me at the moment by defining a class 

```
    using System;
    using System.Runtime.InteropServices;

    internal static class WinUIDesktopInterop
    {
        /// <summary>
        /// Interface from microsoft.ui.xaml.media.dxinterop.h
        /// </summary>
        [ComImport]
        [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
        [Guid("63aad0b8-7c24-40ff-85a8-640d944cc325")]
        public interface ISwapChainPanelNative
        {
            [PreserveSig] uint SetSwapChain([In] IntPtr swapChain);
        }
    }
```

and then casting

```
     WinUIDesktopInterop.ISwapChainPanelNative nativePanel = WinRT.CastExtensions.As<WinUIDesktopInterop.ISwapChainPanelNative>(myWinUISwapChainPanel);
```

**Reply by amerkoleci** [reply]
Last commit https://github.com/amerkoleci/Vortice.Windows/commit/1b769b68387caa2dc6d9bb11664f7a55526b8055 adds new Vortice.WinUI based on ProjectReunion.WinUI, can you test and if it works I will push new release on nuget?

**Reply by TomAtWork** [reply]
Thanks, i will be testing it today or tomorrow.

**Reply by TomAtWork** [reply]
Hello Amer, 
i have tested that it works with a demo application. 

If i reference Vortice.WinUI.dll i can
```
SwapChainPanel panel = ...;
ComObject comObject = new ComObject(panel);

Vortice.WinUI.ISwapChainPanelNative swapchainPanelNative = comObject.QueryInterfaceOrNull<Vortice.WinUI.ISwapChainPanelNative>();
swapchainPanelNative.SetSwapChain(_dxgiSwapChain);
```
I have not tested the other mappings (e.g. ISwapChainPanelNative2).

Thank you very much !!!

**Reply by amerkoleci** [reply]
Glad it works!
